### ISSUE TITLE
Appsettings.json credentials settings

### ISSUE THREAD

**Issue by timhargan** [Cimpress-MCP/serilog-sinks-awscloudwatch#57]
Is there a way to use the AWS credentials settings from my AWS section of my appsettings.json file?
Here is an snippet of my file:

"AWS": {
    "Profile": "Test",
    "ProfilesLocation": "credentials",
    "Region": "us-east-1"
  },
"Serilog": {
    "MinimumLevel": "Information",
    "WriteTo": [
	{
        "Name": "AmazonCloudWatch",
        "Args": {
          "minimumLogEventLevel": "Fatal",
          "logGroupName": "MessagePipline/Applications"
        }
    ],
    "Enrich": [ "FromLogContext", "WithMachineName", "WithThreadId" ],
    "Properties": {
      "Application": "Sample",
      "Environment": "Test"
    }
}

**Comment by wparad** [comment]
As long as the credentials for the AWS are setup BEFORE the logger is initialized, the profile will get pulled in.  If you have some code that isn't working can you share it here?

**Comment by wparad** [comment]
You need to move your code `serviceCollection.AddDefaultAWSOptions(configuration.GetAWSOptions());` above the line `new LoggerFactory()`.

How it works:
* If you don't specify credentials when creating an AWS client, it will pull the current value of the `Config`
* Config works by looking at ENV => Profile => passed in values.
* A new AWS client gets created when the Serilog loads the CloudWatchSink.
* At that point the AWS config must have been set.
* Solution make sure AWS config is set correctly before loading serilog, you can validate this by `Console.WriteLine(AWS.Config)`

There would be no reason for, and ill-advised for Serilog (or worse a custom implementation in this library) to grab the appsettings for AWS and pull them in.  The good news is we don't have to, AWS already does that.  So rather than duplicating their implementation, we should have consume it correctly.

Seeing code, it is clear where the issue is, so I'll close this issue.  If there is a still a problem, feel free to continue to comment.

**Comment by timhargan** [comment]
I moved the loggerFactory section down so that it now looks like this:
// add configured instance of logging
            serviceCollection.AddSingleton(new LoggerFactory()
                .AddConsole()
                .AddSerilog(dispose: true)
                .AddDebug()).AddLogging();


            var file = File.CreateText(Path.Combine(Directory.GetCurrentDirectory(), "self.log"));
            Log.Logger = new LoggerConfiguration()
            .ReadFrom.Configuration(configuration)
            .CreateLogger();

and i still get an exception that it cant find the credentials

**Comment by wparad** [comment]
Okay, I figured it out.  The code you have to load the AWS credentials actually isn't supported by AWS, that is MVC specific, and you actually need AWS to load the credentials.  I'm following the issues at [here](https://github.com/aws/aws-sdk-net/issues/499) but looking at the source code for `serviceCollection.AddDefaultAWSOptions(configuration.GetAWSOptions());`, it doesn't do anything valuable.

If you want to load the credentials by AWS, you need to use a support strategy to do that.  The supported mechanisms are listed here:
https://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/net-dg-config-creds.html

One of the ways listed there, haven't tried would be, to pull the value from the appsettings and stick it into AWS:
`AWSConfigs.AWSProfileName = "profileName";`

**Comment by timhargan** [comment]
I am using this:
https://aws.amazon.com/blogs/developer/configuring-aws-sdk-with-net-core/

**Comment by wparad** [comment]
As I mentioned, that library only supports AWS clients injected MVC and does not support any other library.  You'll need to either inject the credentials into AWS or inject a AWSCloudWatch client into this library.

**Comment by timhargan** [comment]
Its pretty basic functionality of .net core.

For example:
https://pioneercode.com/post/dependency-injection-logging-and-configuration-in-a-dot-net-core-console-app

It just uses:
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
Serilog and all of the other sinks I have come across support this so I am a little confused.

**Comment by wparad** [comment]
(Please do not create extraneous github issues, github is for features or bugs, this was concluded that it is neither.)

TL;DR, you can't do what you are asking, you have three options:
* **Preferred Option** Create a local CloudWatchClient, using your choice of methods and pass it to the library.
* Set the global AWS profile variable, as I stated above, and the library will create the client with those credentials automatically.
* Pass the credentials to the library and the library will create a client for you.

Longer version:
The issue isn't with this library, what you are running into is a limitation of both this library you are using and the Microsoft Logging Framework.  The library you are using passes the AWS profile information to ASP.NET and **NOT** to AWS.Config.  Then, whenever you ask ASP.NET for a AWS client, it creates a client AND uses the AWS Credentials you have passed it.  That means any other mechanism for creating AWS clients will not work, including Amazon itself.  Don't believe me, anywhere in your application write the following code:
```csharp
using (var client = new AmazonS3Client()) 
{
    var request = new GetObjectRequest {BucketName = bucketName, Key = keyName};
    using (var response = client.GetObject(request))  
    using (var responseStream = response.ResponseStream)
    using (var reader = new StreamReader(responseStream))
    {
        var title = response.Metadata["x-amz-meta-title"];
        Console.WriteLine("The object's title is {0}, and body {1}", title, reader.ReadToEnd());
    }
}
```

And you'll also get an exception.

Why doesn't this work?  That's because as I pointed out only MVP has the credentials not AWS.  Why can't Amazon pull from ASP.NET to grab the credentials?  The same reason as this library can't, this library has no idea about the `ServiceCollection`.  It actually is worse.  While we could write some code to pull from the `ServiceCollection` as new feature, it wouldn't work.  Why you ask?  That is a very good question, that is because the Microsoft.Logging framework as well as Serilog also don't know about ASP.NET, so neither of them would be able to get the credentials from the `ServiceCollection` to pass to the CloudWatchClient this library needs.

Does that make sense?

**Comment by timhargan** [comment]
Sorry but why are you referring to asp.net since this is a .net core application and has nothing to do with asp

**Comment by wparad** [comment]
Because:
.NET versus .NET Core
and ASP.NET versus non ASP.NET application.

And the code you wrote is 
`ConfigureServices(IServiceCollection serviceCollection)`
which is from ASP.NET, that is where DI comes from
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/startup?view=aspnetcore-2.1

And amazon here:
See here under *ASP.NET*
https://aws.amazon.com/blogs/developer/configuring-aws-sdk-with-net-core/

**Comment by timhargan** [comment]
Thanks for your help.  Sorry for being so dense.  I got the connection working just went back and put the credentials from my credentials file into the config – seemed like the easiest route ☺
I think I am seeing now is that all I am getting in the cloudwatch log is the message and nothing else from the serilog structured data. Am I missing something for the renderer? I did not specify anything for it.

Thanks, Tim

**Comment by thoean** [comment]
Correct. It simply serializes the message by default. However, you can implement any renderer. One example, is to render JSON. See the wiki for more information - https://github.com/Cimpress-MCP/serilog-sinks-awscloudwatch/wiki/Using-a-JSON-renderer

**Comment by williamgueiros** [comment]
get an AWS exception 'No RegionEndpoint or ServiceURL configured'

 fix that, I set the environment variable AWS_REGION with the region ex: us-east-1
